Update generate() to work like fit() and predict()

mattdangerw opened the pull request:
This updates generate to feel more like fit/predict/evaluate.

Inputs to generate can be a dataset, or raw tensors. Inputs can be preprocessed or not depending on if the model has a preprocessor layer attached.

The preprocessing layer is used to preprocess all inputs before generation.

Fixes #911, #912, #913 and #844 (if we go with this approach).

[Colab with usage](https://colab.research.google.com/gist/mattdangerw/c9fc4cf51196b8741d1815dd01fd85c5/generate-with-datasets.ipynb)

mattdangerw commented:
@fchollet @chenmoneygithub opening up a draft of a refactor for generate. We will probably need some tweaks here, but overall I think we need something like this to keep `generate()` feeling like the other high-level APIs on `keras.Model`.

Long term, we probably factor a lot of code from this into a common generative task class.

fchollet commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -189,12 +190,18 @@ def __init__(
+        # Default compilation
+        self.compile(
+            loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+            optimizer=keras.optimizers.Adam(2e-5),
```
Random note -- I thought AdamW was preferred?

mattdangerw replied:
We have so far avoided it as for tf 2.9 adamw is very broken (does not converge with no error messages). But now that 2.9 is no longer the default version on colab, we could consider using it everywhere.

fchollet commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -291,9 +298,14 @@ def make_generate_function(self):
             return self.generate_function
 
-        def generate_function(prompt, input_mask, min_length):
+        def generate_function(inputs):
```
Please add a comment with the expect structured for `inputs`

mattdangerw replied:
made it flat!

fchollet commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
         Args:
-            prompt: a string, string Tensor or string RaggedTensor. The prompt
-                text for generation.
+            inputs: a list of strings, `tf.Tensor`, or `tf.data.Dataset`.
```
Starting with "a list" is a bit confusing (wrt Tensor and Dataset case)

"a string `tf.Tensor`, a `tf.data.Dataset` of strings, or a list of Python strings"

fchollet commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
-        continue until `max_length` is met, and all tokens generated after
-        `end_token` will be truncated. The sampling approach used can be
-        controlled via the sampler argument.
+        This method generates text based given `inputs` as a prompt.
```
Please clarify that each entry in the batch is treated independently

fchollet commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
-                "`self.preprocessor` is `None`, please make sure "
-                "`preprocessor` is set before calling `generate`."
+                "When `inputs` is a `tf.data.Dataset`, please do not provide "
+                "`batch_size`. Received: "
```
(the Dataset is expected to be batched)

fchollet commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
-                "`preprocessor` is set before calling `generate`."
+                "When `inputs` is a `tf.data.Dataset`, please do not provide "
+                "`batch_size`. Received: "
+                f"`type(batch_size)={type(batch_size)}`."
```
Received batch_size={batch_size} (of type {type(batch_size)})

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
-        output = generate_function(prompt, input_mask, min_length)
+        def preprocess(x, y=None, sample_weight=None):
+            if self.preprocessor is not None:
+                return self.preprocessor(x, sequence_length=max_length)
```
If we use `preprocessor` here, I would vote for setting `add_end_token=False` by default, which is only useful for chatbot model.

mattdangerw replied:
Currently both `add_end_token` and `add_end_token` default to False right?

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_causal_lm_preprocessor.py:
```diff
@@ -74,16 +74,17 @@ class GPT2CausalLMPreprocessor(GPT2Preprocessor):
-
-        x = super().call(x)
+        # Tokenize with one extra token to account for the truncation below.
+        sequence_length = (sequence_length or self.sequence_length) + 1
```
This is a bit odd to me... `sequence_length` has a higher priority than `self.sequence_length` because we need to respect `max_length` in `generate` method, however for users not having this context, this could be weird. Can we override the `sequence_length` in `generate` method?

mattdangerw replied:
like change the name from `max_length` to `sequence_length`? no strong preferences there.

let's avoid mutating config state on a layer, but not sure what the best approach is. this just seemed "least bad" to me

chenmoneygithub reviewed (changes requested):
Thanks Matt! The proposed workflow is nice, and the code is cleaner with `self.packer`! 

Left some comments on implementation details.

mattdangerw commented on keras_nlp/layers/start_end_packer.py:
```diff
@@ -103,48 +109,54 @@ def __init__(
             # Add a new axis at the beginning.
-            inputs = tf.expand_dims(inputs, axis=0)
-        if isinstance(inputs, tf.Tensor):
+            x = tf.expand_dims(x, axis=0)
```
I'm fine with either! You are very correct we are mixed currently. I'll leave as is unless there is preference?

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
-        min_length = tf.reduce_min(prompt.row_lengths())
-        input_mask = tf.ones_like(prompt, tf.bool).to_tensor(shape=padded_shape)
-        prompt = prompt.to_tensor(shape=padded_shape)
+            if batch_size is not None:
```
currently yes, if batch size is passed we use a dataset, and if not we just leave the input as a single batch. we could add some additional speedup code to detect if input cardinality < batch_size

mattdangerw replied:
I think I will leave this as is for now, probably cases where `batch_size` is provided and the input is smaller then a single batch are rare, so not a huge deal. but could be a reasonable follow up

mattdangerw replied:
We could support either a single batch or `tf.data.Dataset` only. That would just mean more discrepancies from `fit()` and `predict()`. For now I was naïvely following the established UX, but I am ok with supporting less options here. Probably worth checking with @fchollet

mattdangerw commented:
/gcbrun

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +340,71 @@ def next(prompt, state, index):
+            inputs = inputs.map(preprocess, tf.data.AUTOTUNE)
+            inputs = inputs.prefetch(tf.data.AUTOTUNE)
+        else:
+            # If we just have a single batch, don't convert to a dataset.
```
Agreed and fixed!

mattdangerw commented on keras_nlp/layers/start_end_packer.py:
```diff
@@ -43,6 +43,11 @@ class StartEndPacker(keras.layers.Layer):
             0 or "" will be added depending on the dtype of the input tensor.
+        return_padding_mask: bool. Weather to return a boolean padding mask of
+            all locations that were filled in with the pad_value. When the
+            `pad_value` occurs naturally in the input or is reused for
```
removed. "reserved" wasn't very clear to me either, so just removed the longer note.

mattdangerw commented:
OK! I have addressed comments and tried to get the high level workflows looking the way we want.

The major awkwardness we are facing is that preprocessing for generation and fine-tuning look quite different. This PR makes all preprocessing run through `model.preprocessor`, and exposes a nice "escape hatch" for generation--by setting `preprocessing=None` on the model, you can run any manipulation of your input and output IDs.

Overall I think this is worth it, but the fact that we are shoving two preprocessing flows into a single task & preprocessor is a little awkward. Options I can think of...

1) Expose call time overrides for the preprocessing layer. `GPT2CausalLMPreprocessor` takes in `sequence_length`, `add_end_token`, `return_labels` at call time to configure the preprocessing. During generation, we pass these arguments to the layer configure the preprocessing the way we want.
2) Split generation into it's own totally separate task. In some ways, this feels very conceptually clean. `GPT2Generator` would not be fittable. `GPT2CausalLM` would be specifically for pre-training/fine-tuning and not have `generate()`. Each task would have a single forward pass. But this would definitely add more labor to an example which first fine-tunes on a dataset, then generates outputs.
3) Expose two separate preprocessing layers on the language model task one for generation and one for training.

Currently I am going with 1., mainly because it is the least disruptive to what we currently have set up.

mattdangerw commented:
/gcbrun

mattdangerw commented:
/gcbrun

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +333,101 @@ def next(prompt, state, index):
+                inputs = tf.data.Dataset.from_tensor_slices(inputs)
+                inputs = inputs.batch(batch_size)
+            else:
+                # If `batch_size` was not provided, just wrap in a list to avoid
```
unsure - should we use "is" or "was" here? usually I go with current tense

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +333,101 @@ def next(prompt, state, index):
+                x,
+                sequence_length=max_length,
+                return_labels=False,
+                add_end_token=False,
```
Let's add a comment here to explain users are expected to do end token padding in string space.

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +333,101 @@ def next(prompt, state, index):
+        outputs = []
+        for batch in inputs:
+            token_ids, mask = batch["token_ids"], batch["padding_mask"]
+            # If preprocessing is attached, we can stop after end_token_id.
```
`preprocessor`

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_preprocessor.py:
```diff
@@ -63,68 +62,63 @@ class GPT2Preprocessor(Preprocessor):
+        sample_weight: Any label weight data. Will be passed through unaltered.
+        sequence_length: Pass to override the configured `sequence_length` of
+            the layer.
+        add_start_token: Pass to override `add_start_token` on the layer.
```
Seems here `add_start_token` should be set as a string, while in `__init__` it is a bool. Shall we avoid saying "override" here?

mattdangerw replied:
I'm confused at this comment, `add_start_token` should never be a string. What do you mean?

chenmoneygithub reviewed (changes requested):
Thanks! The functionality looks good, left some initial comments.

mattdangerw commented on keras_nlp/layers/start_end_packer.py:
```diff
@@ -43,6 +43,17 @@ class StartEndPacker(keras.layers.Layer):
+        inputs: A `tf.Tensor`, `tf.RaggedTensor`, or list of python strings.
+        sequence_length: Pass to override the configured `sequence_length` of
+            the layer.
+        add_start_value: Pass `False` to not append a start value for this
```
Currently `StartEndPacker` can operate in string space or id space. `keras_nlp.layers.StartEndPacker(start_value="start", end_value="end)(some_ragged_string_inputs)` is valid code.

We could remove this functionality (not many use it probably), or leave it supported and undocumented, or leave as is today. But probably best handled off this PR.

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -72,98 +74,84 @@ class GPT2CausalLM(Task):
-        "But I watch youtube while coding!",
-    ]
-    ds = tf.data.Dataset.from_tensor_slices(features).batch(2)
+    # Prompt with 50256, the `"<|endoftext|>"` token id.
```
Oh actually I think example I was showing calling `generate` without preprocessing. So this has nothing to do with our tokenizer.

I updated the example a bit for clarity and remove the endoftext part.

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -72,98 +74,84 @@ class GPT2CausalLM(Task):
+        preprocessor=None,
     )
-    gpt2_lm.predict(["You know this is still a test string"])
+    gpt2_lm.fit(x=x, y=y, sample_weight=sw, batch_size=2)
```
I think we can leave them separate! We are enumerating the main ways the model can be called, but we don't want to show entire workflow here. That is for a guide or example!

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,49 +333,101 @@ def next(prompt, state, index):
         Args:
-            prompt: a string, string Tensor or string RaggedTensor. The prompt
-                text for generation.
+            inputs: a string `tf.Tensor`, a `tf.data.Dataset` of strings, or
```
We do allow it to already! I just didn't want to make the documentation to complex. Update the copy here.

mattdangerw commented:
Comments addressed!

mattdangerw commented:
/gcbrun

mattdangerw commented:
/gcbrun

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+                is attached to the model, inputs should instead be a nested
+                `tf.Tensor` or `tf.data.Dataset` with keys `"token_ids"` and
+                `"padding_mask"`.
+            max_length: int. The max length of the generated sequence.
```
Mark as optional argument and specify behavior is `max_length is None`.

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+
+        outputs = tf.concat(outputs, axis=0)
+        outputs = tf.squeeze(outputs, 0) if input_is_scalar else outputs
+        if self.preprocessor is not None and not return_tensors:
```
Why not just let `return_tensors` control everything? This may be very hard for the user to reason about.

mattdangerw replied:
If `preprocessing` is none, we would be return a list of list of python ints (and the utility below would not work, it would need some reworks). I think the better, keep it simple option would just be to remove `return_tensors`.

If preprocessing -> string output, if not `tf.Tensor` token ids with the same shape that you passed in. WDYT?

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+                inputs = inputs.map(preprocess, tf.data.AUTOTUNE)
+                inputs = inputs.prefetch(tf.data.AUTOTUNE)
+            else:
+                inputs = [preprocess(batch) for batch in inputs]
```
`preprocess` calls `self.preprocessor`, how does this work if `self.preprocessor is None`?

mattdangerw replied:
Note that this is inside a branch `if self.preprocessor is not None:`.

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -291,9 +285,26 @@ def make_generate_function(self):
         if self.generate_function is not None:
```
Couldn't we check whether the method exists at the callsite? Do we need this factory style method?

mattdangerw replied:
We could, but this would be less in line with how `keras.Model` works. See [these lines of keras.Model](https://github.com/keras-team/keras/blob/cbbe8eb53001ccffe9f02d08ecdaf22997879d19/keras/engine/training.py#L1265-L1291). The end goal is to pull this entire function into a base class that feels similar to `keras.Model` in how it handles this stuff.

mattdangerw replied:
Leaving this for now, but happy to keep discussing.

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+                inputs = [preprocess(batch) for batch in inputs]
 
-        # Run the (possibly compiled) generate function on dense inputs.
         generate_function = self.make_generate_function()
```
Just check if `generation_function is None` here.

mattdangerw replied:
See above comment, this is copying the exiting Keras model approach of making a method that "returns or compiles" the underlying function.

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
-        if self.preprocessor is None:
+        input_is_scalar = False
+
+        if isinstance(inputs, tf.data.Dataset) and batch_size is not None:
```
Without the batching functionality we could also kill this

mattdangerw replied:
Yes, if we remove the `batch_size` argument to `generate()`, we could only handle two cases. A dataset or a single batch. I am open to it!

mattdangerw replied:
Removed `batch_size` for now.

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+        outputs will be generated "batch-by-batch" and concatenated. Otherwise,
+        all inputs will be handled in a single generation pass.
+
+        If a `preprocessor` is attached to the model, `inputs` should be
```
You need to discuss `return_tensors` here, depending on the outcome of my questions below

mattdangerw replied:
See above responses!

mattdangerw replied:
Removed return tensors for a follow up PR.

jbischof reviewed (commented):
Thanks! This seems very cool but I've fallen a bit behind on some context. Most of my questions are asking about simplifications.

Also: How much of this code is specific to GPT-2? Would it makes sense to put this in a base class or does this make sense to rewrite from scratch for OPT?

mattdangerw commented:
> Also: How much of this code is specific to GPT-2? Would it makes sense to put this in a base class or does this make sense to rewrite from scratch for OPT?

Definitely a good idea. No way we can support this much complexity for each class.

I've opened https://github.com/keras-team/keras-nlp/issues/868 for it a bit ago. My plan was to keep this as a follow up.

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+                `tf.Tensor` or `tf.data.Dataset` with keys `"token_ids"` and
+                `"padding_mask"`.
+            max_length: int. The max length of the generated sequence.
+            batch_size: int. Only pass if `inputs` is a `tf.Tensor`. If set,
```
The user friendly thing to do is just to throw in that case. `fit()` and `predict()` do not handle batched tensor/numpy input.

If you attempt `bert_classifier.predict(batched_tensors)` you would get an error IIUC. Or any vanilla Keras model for that matter.

jbischof replied:
I see, so one batch at most?

mattdangerw replied:
Well I guess it depend on what you mean by "batched". Do you mean a tensor with shape `(num_batches, batch_size, feature shapes)` or do you mean a python list of tensor batches.

The former would not work in `fit()` or `predict()` etc. The latter might for vanilla Keras? I would need to check.

The main modes of input to the high-level Keras APIs are either a batched dataset, or a single numpy/tensor with shape `(num_sample, feature shapes)` and a `batch_size` provided separately IIUC.

jbischof replied:
It seems that predict has no issues working on batched tensors ([colab](
https://colab.sandbox.google.com/gist/jbischof/27246c9441b06d9bdc3644bf5385f3ff/kerasnlp-predict-batches.ipynb))

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
+            batch_size: int. Only pass if `inputs` is a `tf.Tensor`. If set,
+                `inputs` will be split into `batch_size` batches handled
+                successively.
+            return_tensors: bool. Only used if a `preprocessor` is attached to
```
Maybe the simplest thing to do is to just remove this arg?

This is controlling, for the `preprocessor is not None` case, if we return a `tf.Tensor` of strings, or a python `list` of strings. @fchollet was pointing out on chat that for the strings in strings out case, returning a tensor is somewhat user hostile. If you wanted to further apply tensorflow operations to your output strings, you want `tf.Tensor` output. In almost all other cases you will want a python string.

For the `preprocessing is None` case, it always makes sense to return tensors. You are still going to detokenize yourself, you want to keep things in the format that you can pass to detokenize.

If we don't want to pollute the API, the simplest thing to do is to remove this argument, and always return python strings when preprocessing. Over chat we had discussed adding this argument though.

mattdangerw replied:
Actually remove the arg might break the TFLite flow @chenmoneygithub has been working on, need to check.

mattdangerw replied:
Removed and returning tensors always for now. Will need follow up.

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -72,98 +74,84 @@ class GPT2CausalLM(Task):
+    # Prompt the model with `5338, 318` (the token ids for `"Who is"`).
+    # Use `"padding_mask"` to indicate values that should not be overridden.
+    prompt = {
+        "token_ids": tf.constant([[50256, 0, 0, 0, 0]] * 2),
```
yep, whoops!

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -72,98 +74,84 @@ class GPT2CausalLM(Task):
-    )
+    features = ["The quick brown fox jumped.", "I forgot my homework."]
+    gpt2_lm = keras_nlp.models.GPT2CausalLM.from_preset("gpt2_base_en")
+    gpt2_lm.fit(x=features, batch_size=2)
```
I am worried you might realize that is already part of how `fit()` works for all Keras models in the non-`tf.data` case.

Here is the equivalent line on master for our [BertClassifier](https://github.com/keras-team/keras-nlp/blob/b0c457e9b5ed7c9e48c3707f2c4d2d82bb383759/keras_nlp/models/bert/bert_classifier.py#L103), this is not new!

Understanding inputs to Keras models is a dense and complex topic. If you have time for code splunking sometime, the [data adapter](https://github.com/keras-team/keras/blob/master/keras/engine/data_adapter.py) class in core Keras is worth a dive.

mattdangerw replied:
Leaving this, as it is how we are documenting this for all task level `fit` call currently. Can address any changes as a follow up.

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -189,12 +177,18 @@ def __init__(
-        self._sampler = get_sampler("top_k")
+        self._sampler = None
+
+        # Default compilation
```
Hah, yes this PR could easily be split a lot. I have been checking with @chenmoneygithub and we just kept it a big ol thing, but if you would like me to split it out let me know!

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -324,47 +335,109 @@ def next(prompt, state, index):
-        prompt,
-        max_length,
+        inputs,
+        max_length=None,
```
Sorry for the back and forth, but I am starting to feel give `max_length` a default to model capacity could lead to a bad UX. The vanilla GPT2 won't stop until 1024, so users would need to wait for a while to get results back. But anyway let's get this PR in, and I can open a few followups.

mattdangerw replied:
Let's chat! I think we should think a little more generally and long term. How many people actually want exactly 50 unconditioned tokens for GPT2. It's kinda toy I think?

Most real workflows will usually involve fine-tuning on sequences that terminate in someway. E.g. build a summarizer, or a chatbot responder, or a translator. This get's even more true when we think about seq2seq models like T5 and BART, which will follow the same UX we are establishing here.

We want to choose a default that will scale well towards the future states of the library, and I am skeptical that requiring a `max_length` in all cases is a good idea.

chenmoneygithub commented on keras_nlp/models/gpt2/gpt2_preprocessor.py:
```diff
@@ -63,68 +62,65 @@ class GPT2Preprocessor(Preprocessor):
+        sample_weight: Any label weight data. Will be passed through unaltered.
+        sequence_length: Pass to override the configured `sequence_length` of
+            the layer.
+        add_start_token: Pass to override the configure value of
```
(cannot remember if I have posted this or not...)

So in call and __init__, we both have `add_start_token`, but they are actually different types. This could lead to some confusion.

mattdangerw replied:
There was a comment chain above that confused me, but basically if we go with this override approach, we need three possible types in call so it can act as an actual override. `None` default use configured value, `True` override to true, `False` override to false. If you only accept True/False here, you have made the init argument obsolete right?

chenmoneygithub reviewed (approved):
Approved! 

Dropped some nonblocking comments, let's keep discussing and do roll forward fixes.

mattdangerw commented:
/gcbrun

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -340,47 +352,87 @@ def next(prompt, cache, index):
-        continue until `max_length` is met, and all tokens generated after
-        `end_token` will be truncated. The sampling strategy can be set in
-        the `compile` method.
+        This method generates text based on given `prompt`. The sampling used
```
It looks like `prompt` has been migrated to `inputs`

mattdangerw replied:
It felt weird to call something that could be a dataset or a tuples of tensors including a mask a `prompt`. WDYT?

jbischof replied:
It's fine to use prompt as a term but backtick `prompt` usually means an variable name. Maybe say based on the prompts in `inputs`?

mattdangerw replied:
Oh got it! Misunderstood the comment. Will just use `inputs` everywhere for now.

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -340,47 +352,87 @@ def next(prompt, cache, index):
-                "`self.preprocessor` is `None`, please make sure "
-                "`preprocessor` is set before calling `generate`."
-            )
+        If `inputs` are a batched `tf.data.Dataset` or `batch_size` is provided,
```
Remove discussion of `batch_size`

jbischof commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -340,47 +352,87 @@ def next(prompt, cache, index):
+                is attached to the model, inputs should instead be a nested
+                `tf.Tensor` or `tf.data.Dataset` with keys `"token_ids"` and
+                `"padding_mask"`.
+            max_length: Optional. int. The max length of the generated sequence.
```
But what if there's no preprocessor? In that case this arg appears not to work.

mattdangerw replied:
Yeah, this arg is worth discussing. `max_length` is really conceptually a preprocessing arg and does absolutely nothing if we are passing preprocessed inputs, which should already have shape `(batch_size, max_length)`.

This feels like what you would call a denomalized argument somewhat. In that `sequence_length` on the preprocessor and `max_length` here are really setting the same thing.

We could...
 - Think about removing it, though I would do that as a follow perhaps.
 - Document that it does nothing when `preprocessor is None`.
 - Throw an error if `inputs.shape[1] != max_length`.

Thoughts?

mattdangerw replied:
I think I will go with the lightweight approach of documenting for now.

jbischof replied:
Sight edit for clarity:

```python
"""
If `preprocessor` is `None`, `inputs` should be padded to the desired maximum 
length and this argument will be ignored.
"""
```

jbischof reviewed (commented):
Seems reasonable just a few questions

mattdangerw commented on keras_nlp/models/gpt2/gpt2_causal_lm.py:
```diff
@@ -72,98 +73,84 @@ class GPT2CausalLM(Task):
+        backbone=backbone,
+        preprocessor=preprocessor,
     )
+    gpt2_lm.generate("I want to say", max_length=30)
```
Good call. I am going to switch to showing `fit` for this bring your own data example. Otherwise we would be showing generate on a random backbone, which would have garbage results anyway.

mattdangerw commented:
/gcbrun

jbischof reviewed (approved):
Thanks for the hard work here!

mattdangerw commented:
/gcbrun
